File: fastapi_server/app/models/user_room_recommendation.py
from sqlalchemy import Column, Float, ForeignKey, Integer, UniqueConstraint
from sqlalchemy.orm import relationship
from database import Base

class UserRoomRecommendation(Base):
    __tablename__ = "user_room_recommendations"
    __table_args__ = (
        UniqueConstraint("user_id", "room_id", name="uq_user_room"),
    )

    user_room_rec_id = Column(Integer, primary_key=True, index=True, autoincrement=True)
    user_id = Column(Integer, ForeignKey("users.user_id"), nullable=False)
    room_id = Column(Integer, ForeignKey("rooms.room_id"), nullable=False)
    score = Column(Float, nullable=False)
    activity_score = Column(Float, nullable=False)

    user = relationship("User")
    room = relationship("Room")

File: spring_server/src/main/java/com/semothon/spring_server/room/entity/UserRoomRecommendation.java
package com.semothon.spring_server.room.entity;

import com.semothon.spring_server.user.entity.User;
import jakarta.persistence.*;
import lombok.*;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
@ToString(of = {"userRoomRecId", "score", "activityScore"})
@Table(name = "user_room_recommendations",
        indexes = {
        },
        uniqueConstraints = {
                @UniqueConstraint(columnNames = {"room_id", "user_id"})
        }
)
public class UserRoomRecommendation {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long userRoomRecId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "room_id", nullable = false)
    private Room room;

    @Column(nullable = false)
    private Double score;

    @Column(nullable = false)
    private Double activityScore;

    public void updateUser(User user){
        this.user = user;
        user.addUserRoomRecommendation(this);
    }

    public void updateRoom(Room room){
        this.room = room;
        room.addUserRoomRecommendation(this);
    }

    public void updateActivityScore(Double score) {
        this.activityScore = score;
    }

}
